### PR TITLE
New Asset System - Add Sub Asset Classes and extended Market functionality

### PR DESCRIPTION
**Pull request by sea212**
### What does it do?
- Add sub asset classes `BaseAssetClass`, `ParimutuelAssetClass`, `XcmAssetClass`
  - This allows specifying proper types for extrinsics (i.e. not allowing `ForeignAsset` as outcome in parimutuels or `MarketAsset` as base assets in markets)
- Extend market struct functionality (e.g. get outcomes)
- Provide `StateTransitionApi` that is used to provide all pallets the option to hook into a market state transition (this will be necessary to delegate market asset creation / destruction)

### What important points should reviewers know?
- This is a part of https://github.com/zeitgeistpm/zeitgeist/pull/1260, which is split into four separate PRs

### CODE REVIEW

File: primitives/src/assets/subsets/parimutuel.rs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Zeitgeist. If not, see <https://www.gnu.org/licenses/>.
+
+use super::*;
+
+/// The `BaseAssets` enum represents all assets that can be used as collateral in
+/// prediction markets.
```
Comment (maltekliemann) [inline primitives/src/assets/subsets/parimutuel.rs:21]: Wrong docs here.

File: primitives/src/assets/tests/conversion.rs
Comment (maltekliemann) [inline primitives/src/assets/tests/conversion.rs:1]: We should also test the failure paths of `try_into`. For example, trying to convert a `ForeignAsset` into a `ParimutuelShare` should fail.

File: primitives/src/market.rs [hunk 2]
```diff
@@ -79,6 +79,105 @@ impl<AI, BA, BN, M, A> Market<AI, BA, BN, M, A> {
     pub fn is_redeemable(&self) -> bool {
         matches!(self.resolution_mechanism(), ResolutionMechanism::RedeemTokens)
     }
+
```
Comment (maltekliemann) [inline primitives/src/market.rs:82]: We should add documentation to these functions.

File: zrml/asset-router/src/pallet_impl/managed_destroy.rs [hunk 3]
```diff
@@ -58,7 +58,6 @@ impl<T: Config> ManagedDestroy<T::AccountId> for Pallet<T> {
         Ok(())
     }
 
-    #[require_transactional]
```
[the remark this answers is not in the document]
Comment (sea212) [reply zrml/asset-router/src/pallet_impl/managed_destroy.rs:61]: This function is called within `MarketTransitionApi` impl of `Parimutuels` which I initially also market as `require_transactional`. Unfortunately it seems to be impossible to return `ResultWithWeightInfo` (or anything other than `DispatchResult`) within functions that `require_transactional`, as it expects `DispatchResult`.

File: primitives/src/market.rs
[the remark this answers is not in the document]
Comment (sea212) [reply primitives/src/market.rs:1]: What exactly are you referring to? 
I regards to the new functionality I added, all execution paths are covered by the tests `provides_correct_list_of_assets` and `converts_outcome_correctly`.

File: primitives/src/constants/mock.rs [hunk 4]
```diff
@@ -189,7 +189,7 @@ parameter_type_with_key! {
 
 parameter_type_with_key! {
     pub ExistentialDepositsNew: |_currency_id: Currencies| -> Balance {
-        1
+        2
```
[the remark this answers is not in the document]
Comment (sea212) [reply primitives/src/constants/mock.rs:192]: Some tests (in orderbook specifically) test the case that the resulting balance is greater than `0` but less than `ExistentialDeposit`. This is not possible if `ExistentialDeposit = 1`.

File: primitives/src/assets/subsets/parimutuel.rs (hunk 1 above)
Comment (sea212) [reply primitives/src/assets/subsets/parimutuel.rs]: https://github.com/zeitgeistpm/zeitgeist/pull/1281/commits/a5fa9b8fbdd2763365440f0400c0c92ae26e956a

File: primitives/src/market.rs (hunk 2 above)
Comment (sea212) [reply primitives/src/market.rs:88]: https://github.com/zeitgeistpm/zeitgeist/pull/1281/commits/1a1768fd3ea64242c41e19a48408b930800d0cf3

File: primitives/src/traits/market_transition_api.rs
[the remark this answers is not in the document]
Comment (sea212) [reply primitives/src/traits/market_transition_api.rs:1]: https://github.com/zeitgeistpm/zeitgeist/pull/1281/commits/e7563f7a4dfbec3249fcd42705ca5fd798cc1136

File: primitives/src/assets/tests/conversion.rs
Comment (sea212) [reply primitives/src/assets/tests/conversion.rs:1]: https://github.com/zeitgeistpm/zeitgeist/pull/1281/commits/2020bb89554c634bc6cf56197c36f1aebc98274c

File: primitives/src/market.rs
[the remark this answers is not in the document]
Comment (maltekliemann) [reply primitives/src/market.rs:1]: Sorry, must have missed some of those.

File: primitives/src/traits/market_transition_api.rs [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+        }
+        fn on_dispute(_market_id: &u128) -> ResultWithWeightInfo<DispatchResult> {
+            panic!("on_dispute");
+        }
+        fn on_resolution(_market_id: &u128) -> ResultWithWeightInfo<DispatchResult> {
+            panic!("on_resolution");
+        }
+    }
+
+    struct SucessPath;
```
Comment (maltekliemann) [inline primitives/src/traits/market_transition_api.rs:153]: Should be called `SuccessPath`.

**Review by maltekliemann — APPROVED** [review]

File: primitives/src/market.rs [hunk 6]
```diff
@@ -69,16 +69,121 @@ pub struct Market<AI, BA, BN, M, A> {
...
+                    Some(MarketAssetClass::<MI>::CategoricalOutcome(market_id, *idx))
+                }
+                ScoringRule::Lmsr => {
+                    Some(MarketAssetClass::<MI>::CategoricalOutcome(market_id, *idx))
+                }
+                ScoringRule::Parimutuel => {
+                    Some(MarketAssetClass::<MI>::ParimutuelShare(market_id, *idx))
+                }
+            },
+            OutcomeReport::Scalar(_) => None,
```
Comment (maltekliemann) [inline primitives/src/market.rs:184]: Actually, one more thing. I'm kind of confused by this one. Why do scalar markets return `None`? `report_into_asset` is not documented, so I'm not sure if this is expected behavior, but it sure is confusing.

**Review by maltekliemann — CHANGES_REQUESTED** [review]

File: primitives/src/traits/market_transition_api.rs (hunk 5 above)
Comment (sea212) [reply primitives/src/traits/market_transition_api.rs]: [SuccessPath](https://github.com/zeitgeistpm/zeitgeist/pull/1281/commits/427b7cedde5a95c2eb002ed3684678aa8f0812f2)

File: primitives/src/market.rs (hunk 6 above)
Comment (sea212) [reply primitives/src/market.rs:195]: Added docstrings for the new public functions: https://github.com/zeitgeistpm/zeitgeist/pull/1281/commits/de8944a535baba0bd60e89018d5b0afa5f56c7c9

The reason it returns `None` is that no asset exists that represents the `OutcomeReport`. To be more precise, the `OutcomeReport::Scalar(u128)` provides a precise number of the outcome, whereas the underlying assets used for trading are `MarketAsset::ScalarOutcome(id, Long/Short)`.

**Review by maltekliemann — APPROVED** [review]

**Pull request merged**